From 199eb5140f12d71c62427703c0011c6d63c03b74 Mon Sep 17 00:00:00 2001
From: Lukas Rusak <lorusak@gmail.com>
Date: Wed, 5 Sep 2018 23:14:09 -0700
Subject: [PATCH] gbm: fix surface creation when no modifiers are present

DRM_MOD_FORMAT_LINEAR should always be present so the modifier count
should always be > 0. Before the surface would get created but was
corrupted. It seems like odd behaviour but this now matches what
kmscube is doing for surface creation.
---
 xbmc/windowing/gbm/GBMUtils.cpp     | 14 ++++++++------
 xbmc/windowing/gbm/WinSystemGbm.cpp |  2 +-
 2 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/xbmc/windowing/gbm/GBMUtils.cpp b/xbmc/windowing/gbm/GBMUtils.cpp
index 118b99f94769..a48d6cc4f84a 100644
--- a/xbmc/windowing/gbm/GBMUtils.cpp
+++ b/xbmc/windowing/gbm/GBMUtils.cpp
@@ -48,13 +48,15 @@ bool CGBMUtils::CreateSurface(int width, int height, const uint64_t *modifiers,
                                                 GBM_FORMAT_ARGB8888,
                                                 modifiers,
                                                 modifiers_count);
-#else
-  m_surface = gbm_surface_create(m_device,
-                                 width,
-                                 height,
-                                 GBM_FORMAT_ARGB8888,
-                                 GBM_BO_USE_SCANOUT | GBM_BO_USE_RENDERING);
 #endif
+  if (!m_surface)
+  {
+    m_surface = gbm_surface_create(m_device,
+                                   width,
+                                   height,
+                                   GBM_FORMAT_ARGB8888,
+                                   GBM_BO_USE_SCANOUT | GBM_BO_USE_RENDERING);
+  }
 
   if (!m_surface)
   {
diff --git a/xbmc/windowing/gbm/WinSystemGbm.cpp b/xbmc/windowing/gbm/WinSystemGbm.cpp
index 9484f5aef4e2..a01746673ec4 100644
--- a/xbmc/windowing/gbm/WinSystemGbm.cpp
+++ b/xbmc/windowing/gbm/WinSystemGbm.cpp
@@ -129,7 +129,7 @@ bool CWinSystemGbm::CreateNewWindow(const std::string& name,
 
   std::vector<uint64_t> *modifiers = m_DRM->GetOverlayPlaneModifiersForFormat(m_DRM->GetOverlayPlane()->format);
 
-  if (!m_GBM->CreateSurface(res.iWidth, res.iHeight, modifiers->data(), modifiers->size()))
+  if (!m_GBM->CreateSurface(res.iWidth, res.iHeight, modifiers->data(), modifiers->size() == 0 ? 1 : modifiers->size()))
   {
     CLog::Log(LOGERROR, "CWinSystemGbm::%s - failed to initialize GBM", __FUNCTION__);
     return false;
